### PR TITLE
Open-API: Refactor TableRequirements

### PR DESCRIPTION
**Pull request by Fokko**
Noticed this when working on https://github.com/apache/iceberg/pull/6323

I believe the requirements are not defined in the right way. After generating code I got:

```python
class TableRequirement(BaseModel):
    type: Literal[
        'assert-create',
        'assert-table-uuid',
        'assert-ref-snapshot-id',
        'assert-last-assigned-field-id',
        'assert-current-schema-id',
        'assert-last-assigned-partition-id',
        'assert-default-spec-id',
        'assert-default-sort-order-id',
    ]
    ref: Optional[str] = None
    uuid: Optional[str] = None
    snapshot_id: Optional[int] = Field(None, alias='snapshot-id')
    last_assigned_field_id: Optional[int] = Field(None, alias='last-assigned-field-id')
    current_schema_id: Optional[int] = Field(None, alias='current-schema-id')
    last_assigned_partition_id: Optional[int] = Field(
        None, alias='last-assigned-partition-id'
    )
    default_spec_id: Optional[int] = Field(None, alias='default-spec-id')
    default_sort_order_id: Optional[int] = Field(None, alias='default-sort-order-id')
```

Which encapsulates all the requirements. After the refactor in this PR, we'll end up with:

```python
class TableRequirement(BaseModel):
    type: str


class AssertCreate(TableRequirement):
    """
    The table must not already exist; used for create transactions
    """

    type: Literal['assert-create']


class AssertTableUUID(TableRequirement):
    """
    The table UUID must match the requirement's `uuid`
    """

    type: Literal['assert-table-uuid']
    uuid: str


class AssertRefSnapshotId(TableRequirement):
    """
    The table branch or tag identified by the requirement's `ref` must reference the requirement's `snapshot-id`; if `snapshot-id` is `null` or missing, the ref must not already exist
    """

    type: Literal['assert-ref-snapshot-id']
    ref: str
    snapshot_id: int = Field(..., alias='snapshot-id')


class AssertLastAssignedFieldId(TableRequirement):
    """
    The table's last assigned column id must match the requirement's `last-assigned-field-id`
    """

    type: Literal['assert-last-assigned-field-id']
    last_assigned_field_id: int = Field(..., alias='last-assigned-field-id')


class AssertCurrentSchemaId(TableRequirement):
    """
    The table's current schema id must match the requirement's `current-schema-id`
    """

    type: Literal['assert-current-schema-id']
    current_schema_id: int = Field(..., alias='current-schema-id')


class AssertLastAssignedPartitionId(TableRequirement):
    """
    The table's last assigned partition id must match the requirement's `last-assigned-partition-id`
    """

    type: Literal['assert-last-assigned-partition-id']
    last_assigned_partition_id: int = Field(..., alias='last-assigned-partition-id')


class AssertDefaultSpecId(TableRequirement):
    """
    The table's default spec id must match the requirement's `default-spec-id`
    """

    type: Literal['assert-default-spec-id']
    default_spec_id: int = Field(..., alias='default-spec-id')


class AssertDefaultSortOrderId(TableRequirement):
    """
    The table's default sort order id must match the requirement's `default-sort-order-id`
    """

    type: Literal['assert-default-sort-order-id']
    default_sort_order_id: int = Field(..., alias='default-sort-order-id')
```

Which makes sense to me.

See below for the final version.

### CODE REVIEW

File: open-api/rest-catalog-open-api.yaml [hunk 1]
```diff
@@ -1649,58 +1649,134 @@ components:
...
           type: integer
           format: int64
-        last-assigned-field-id:
+
+    AssertLastAssignedFieldId:
+      description:
+        The table's last assigned column id must match the requirement's `last-assigned-field-id`
+      required:
+        - type
+        - last-assigned-partition-id
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:1697]:
```suggestion
        - last-assigned-field-id
```

File: open-api/rest-catalog-open-api.yaml [hunk 2]
```diff
@@ -1649,58 +1649,134 @@ components:
...
+      properties:
+        type:
+          type: string
+          enum: [ "assert-last-assigned-field-id" ]
+        last-assigned-partition-id:
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:1702]:
```suggestion
        last-assigned-field-id:
```

File: open-api/rest-catalog-open-api.yaml [hunk 3]
```diff
@@ -1649,58 +1649,134 @@ components:
...
+        - default-sort-order-id
+      properties:
+        type:
+          type: string
+          enum: [ "assert-default-sort-order-id" ]
         default-sort-order-id:
           type: integer
 
+    TableRequirement:
+      oneOf:
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:1758]: would it make sense to define the `type` enum here rather than having it in every subtype? (similar to how we do it for `    ReportMetricsRequest`)

Comment (Fokko) [reply open-api/rest-catalog-open-api.yaml:1758]: I don't think the `ReportMetricsRequest` metric is correct.

```yaml
    ReportMetricsRequest:
      anyOf:
        - $ref: '#/components/schemas/ScanReport'
        - $ref: '#/components/schemas/CommitReport'
      required:
        - report-type
      properties:
        report-type:
          type: string
```

Let's take the `report-type`. Which value corresponds to the `ScanReport` or `CommitReport`? This is `scan-report` and `commit-report`, but this is not described in the spec. If you would generate code from it, and you have a class that accepts a `ReportMetricRequests`:

```yaml
    post:
      tags:
        - Catalog API
      summary: Send a metrics report to this endpoint to be processed by the backend
      operationId: reportMetrics
      requestBody:
        description: The request containing the metrics report to be sent
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/ReportMetricsRequest'
```

It cannot determine if is a `ScanReport` or `CommitReport`, this is where you need the discriminator.

**Comment by Fokko** [comment]
Alright, now I have it right:

```python
class TableRequirement(BaseModel):
    name: Optional[str] = None


class AssertCreate(TableRequirement):
    """
    The table must not already exist; used for create transactions
    """

    type: Literal['assert-create']


class AssertTableUUID(TableRequirement):
    """
    The table UUID must match the requirement's `uuid`
    """

    type: Literal['assert-table-uuid']
    uuid: str


class AssertRefSnapshotId(TableRequirement):
    """
    The table branch or tag identified by the requirement's `ref` must reference the requirement's `snapshot-id`; if `snapshot-id` is `null` or missing, the ref must not already exist
    """

    type: Literal['assert-ref-snapshot-id']
    ref: str
    snapshot_id: int = Field(..., alias='snapshot-id')


class AssertLastAssignedFieldId(TableRequirement):
    """
    The table's last assigned column id must match the requirement's `last-assigned-field-id`
    """

    type: Literal['assert-last-assigned-field-id']
    last_assigned_field_id: int = Field(..., alias='last-assigned-field-id')


class AssertCurrentSchemaId(TableRequirement):
    """
    The table's current schema id must match the requirement's `current-schema-id`
    """

    type: Literal['assert-current-schema-id']
    current_schema_id: int = Field(..., alias='current-schema-id')


class AssertLastAssignedPartitionId(TableRequirement):
    """
    The table's last assigned partition id must match the requirement's `last-assigned-partition-id`
    """

    type: Literal['assert-last-assigned-partition-id']
    last_assigned_partition_id: int = Field(..., alias='last-assigned-partition-id')


class AssertDefaultSpecId(TableRequirement):
    """
    The table's default spec id must match the requirement's `default-spec-id`
    """

    type: Literal['assert-default-spec-id']
    default_spec_id: int = Field(..., alias='default-spec-id')


class AssertDefaultSortOrderId(TableRequirement):
    """
    The table's default sort order id must match the requirement's `default-sort-order-id`
    """

    type: Literal['assert-default-sort-order-id']
    default_sort_order_id: int = Field(..., alias='default-sort-order-id')
```

The inheritance tree now checks out.

File: open-api/rest-catalog-open-api.yaml [hunk 4]
```diff
@@ -1824,58 +1824,146 @@ components:
         - $ref: '#/components/schemas/SetPropertiesUpdate'
         - $ref: '#/components/schemas/RemovePropertiesUpdate'
 
-    TableRequirement:
```
Comment (rdblue) [inline open-api/rest-catalog-open-api.yaml:1827]: Nit: Should TableRequirement be defined first?

**Review by rdblue — APPROVED** [review]
This looks good to me. I like using the discriminator. We should probably refactor other areas to have that structure.

**Comment by rdblue** [comment]
Sorry for the late review. This looks great. Thanks @Fokko!

I didn't merge this in case it conflicts with the other linked PR. Feel free to merge in the order that makes the most sense.

File: open-api/rest-catalog-open-api.yaml (hunk 4 above)
Comment (Fokko) [reply open-api/rest-catalog-open-api.yaml:1827]: Yes, I don't recall this isn't the case.

**Comment by Fokko** [comment]
Thanks for the review @rdblue and @nastra for the review. I'll might revisitor other parts as well 👍

**Comment by romusz** [comment]
Hi,

Have you tested this refactoring with code generated for Go?
What we get now is:
```go
type TableRequirement struct {
	Type string `json:"type"`
}
```

and (for example)

```go
type AssertTableUUID struct {
	TableRequirement
	Type string `json:"type"`
	Uuid string `json:"uuid"`
}
```

There are couple of problems with that:
1. double storage of `Type`
2. provided `[]TableRequirement{}` slice and the fact that `TableRequirement` is a struct not an interface, it cannot be "type switched" into it's "subtypes" (e.g. `AssertTableUUID`). 
3. All previously-available, useful methods on `TableRequirement` have been nuked (e.g. `GetUuidOk` that allowed to extract TableUUID from `assert-table-uuid` requirement).

In summary, Go generated code does not produce equivalent to what it's shown for Python.
Please let me know if there is other way to traverse the requirements - so far it looks like a regression.

**Comment by romusz** [comment]
@Fokko not sure if you'll see the above comment since the pr has been merged, thus an explicit call-out.
